feature/subscriber-attributes/src/main/java/com/revenuecat/purchases/subscriberattributes/SubscriberAttributesManager.kt

```diff
@@ -160,8 +173,49 @@ class SubscriberAttributesManager(
         applicationContext: Application,
         completion: (deviceIdentifiers: Map<String, String?>) -> Unit
     ) {
+        obtainingDeviceIdentifiersObservable.numberOfProcesses++
         deviceIdentifiersFetcher.getDeviceIdentifiers(applicationContext) { deviceIdentifiers ->
             completion(deviceIdentifiers)
+            obtainingDeviceIdentifiersObservable.numberOfProcesses--
+        }
+    }
+
+    // This was added to resolve: https://revenuecats.atlassian.net/browse/CSDK-433. To summarize it,
+    // setting advertising IDs happens on a non-main thread. If that process is delayed for any reason, the attributes
+    // won't be stored in the device cache in time for the sync with the backend. This is a mechanism added so we can
+    // wait until the device identifiers have been obtained before continuing with other processes.
+    private class ObtainDeviceIdentifiersObservable : Observable() {
+        var numberOfProcesses: Int = 0
+            @Synchronized
+            get
+            @Synchronized
+            set(value) {
+                if (field == value) return
+                field = value
+                setChanged()
+                notifyObservers()
+            }
+
+        private val listeners: ArrayList<() -> Unit> = arrayListOf()
+            @Synchronized
+            get
```

Comment:
You can also do this which I think is cleaner:

```suggestion
        @get:Synchronized
        private val listeners: ArrayList<() -> Unit> = arrayListOf()
```